# Набор данных

**Оценка стоимости  и типа трансмиссии по данным продаж автомобилей на вторичном рынке Молдавии**

Набор данных представляет собой статистику параметров автомобилей на вторичном рынке в Молдавии. 

Набор включает ряд категориальных и численных значений, составляющих одну запись (строку). Число записей можно найти как число строк.

Каждый столбец в записи — это отдельный параметр. 

Среди указанных параметров приведены целевой для задачи предсказания (регрессии) - цена автомобиля.

 Также среди параметров есть целевой для задачи классификации - тип трансмиссии. 
 
 Последняя задача может быть рассмотрена, например, как пример задачи на заполнение пропусков (если продавец не указал соответствующий параметр).

# Import библиотек

In [40]:
import pandas as pd
from google.colab import files 

# Загрузка данных на Google Drive

In [42]:
uploaded = files.upload()

Saving cars_moldova.csv to cars_moldova (3).csv


# Считываем файл в DataFrame



метод .read_csv(path,delimiter) 

In [53]:
df = pd.read_csv('/content/cars_moldova (3).csv', delimiter = ',')# мне пришлось менять название файла, потому что с обычным он не хотел работать нормально

In [44]:
df

Make         Model  Year      Style  Distance  \
0          Toyota         Prius  2011  Hatchback  195000.0   
1         Renault  Grand Scenic  2014  Universal  135000.0   
2      Volkswagen          Golf  1998  Hatchback       1.0   
3         Renault        Laguna  2012  Universal  110000.0   
4            Opel         Astra  2006  Universal  200000.0   
...           ...           ...   ...        ...       ...   
41002       Dacia     Logan Mcv  2015  Universal   89000.0   
41003     Renault         Modus  2009  Hatchback     225.0   
41004    Mercedes       E Class  2016      Sedan   50000.0   
41005       Mazda             6  2006      Combi  370000.0   
41006     Renault  Grand Scenic  2006    Minivan  300000.0   

       Engine_capacity(cm3)     Fuel_type Transmission  Price(euro)  
0                    1800.0        Hybrid    Automatic       7750.0  
1                    1500.0        Diesel       Manual       8550.0  
2                    1400.0        Petrol       Manual       2200.0  
3                    1500.0        Diesel       Manual       6550.0  
4                    1600.0  Metan/Propan       Manual       4100.0  
...                     ...           ...          ...          ...  
41002                1500.0        Diesel       Manual       7000.0  
41003                1500.0        Diesel       Manual       4500.0  
41004                1950.0        Diesel    Automatic      29500.0  
41005                2000.0        Diesel       Manual       4000.0  
41006                1500.0        Diesel       Manual       4000.0  

[41007 rows x 9 columns]

# Информация о DataFrame


Метод .info() 

In [47]:
df.info() # считываем информацию о списке 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41007 entries, 0 to 41006
Data columns (total 9 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Make                  41007 non-null  object 
 1   Model                 41007 non-null  object 
 2   Year                  41007 non-null  int64  
 3   Style                 41007 non-null  object 
 4   Distance              41007 non-null  float64
 5   Engine_capacity(cm3)  41007 non-null  float64
 6   Fuel_type             41007 non-null  object 
 7   Transmission          41007 non-null  object 
 8   Price(euro)           41007 non-null  float64
dtypes: float64(3), int64(1), object(5)
memory usage: 2.8+ MB


# Поиск Дубликатов
 


методы .duplicated() и .sum()

In [48]:
df.duplicated().sum() # вычисляем сколько всего дубликатов 

3743

**Удаление дубликатов**

Методы .drop_duplicates() и .reset_index()

In [49]:
DF = df.drop_duplicates()# убиравем дубликаты 

DF = DF.reset_index(drop=True)# делаем список обновляя индексацию с убранными дубликатами  и выводим его 

DF

Make         Model  Year      Style  Distance  \
0          Toyota         Prius  2011  Hatchback  195000.0   
1         Renault  Grand Scenic  2014  Universal  135000.0   
2      Volkswagen          Golf  1998  Hatchback       1.0   
3         Renault        Laguna  2012  Universal  110000.0   
4            Opel         Astra  2006  Universal  200000.0   
...           ...           ...   ...        ...       ...   
37259  Land Rover    Freelander  2002  Crossover  225000.0   
37260       Dacia     Logan Mcv  2015  Universal   89000.0   
37261     Renault         Modus  2009  Hatchback     225.0   
37262       Mazda             6  2006      Combi  370000.0   
37263     Renault  Grand Scenic  2006    Minivan  300000.0   

       Engine_capacity(cm3)     Fuel_type Transmission  Price(euro)  
0                    1800.0        Hybrid    Automatic       7750.0  
1                    1500.0        Diesel       Manual       8550.0  
2                    1400.0        Petrol       Manual       2200.0  
3                    1500.0        Diesel       Manual       6550.0  
4                    1600.0  Metan/Propan       Manual       4100.0  
...                     ...           ...          ...          ...  
37259                1800.0  Metan/Propan       Manual       4400.0  
37260                1500.0        Diesel       Manual       7000.0  
37261                1500.0        Diesel       Manual       4500.0  
37262                2000.0        Diesel       Manual       4000.0  
37263                1500.0        Diesel       Manual       4000.0  

[37264 rows x 9 columns]

In [50]:
DF.to_csv('/content/cars_moldova_no_dup.csv',index=False) #. сохраняем все в файл 

In [51]:
DF

Make         Model  Year      Style  Distance  \
0          Toyota         Prius  2011  Hatchback  195000.0   
1         Renault  Grand Scenic  2014  Universal  135000.0   
2      Volkswagen          Golf  1998  Hatchback       1.0   
3         Renault        Laguna  2012  Universal  110000.0   
4            Opel         Astra  2006  Universal  200000.0   
...           ...           ...   ...        ...       ...   
37259  Land Rover    Freelander  2002  Crossover  225000.0   
37260       Dacia     Logan Mcv  2015  Universal   89000.0   
37261     Renault         Modus  2009  Hatchback     225.0   
37262       Mazda             6  2006      Combi  370000.0   
37263     Renault  Grand Scenic  2006    Minivan  300000.0   

       Engine_capacity(cm3)     Fuel_type Transmission  Price(euro)  
0                    1800.0        Hybrid    Automatic       7750.0  
1                    1500.0        Diesel       Manual       8550.0  
2                    1400.0        Petrol       Manual       2200.0  
3                    1500.0        Diesel       Manual       6550.0  
4                    1600.0  Metan/Propan       Manual       4100.0  
...                     ...           ...          ...          ...  
37259                1800.0  Metan/Propan       Manual       4400.0  
37260                1500.0        Diesel       Manual       7000.0  
37261                1500.0        Diesel       Manual       4500.0  
37262                2000.0        Diesel       Manual       4000.0  
37263                1500.0        Diesel       Manual       4000.0  

[37264 rows x 9 columns]

# Индексация

**Визуализация части DataFrame**

Метод .head(n) - первые n строк

In [ ]:
DF.head(6) #. показываем первые строки (кол-во строк в скобках)

Make         Model  Year      Style  Distance  Engine_capacity(cm3)  \
0      Toyota         Prius  2011  Hatchback  195000.0                1800.0   
1     Renault  Grand Scenic  2014  Universal  135000.0                1500.0   
2  Volkswagen          Golf  1998  Hatchback       1.0                1400.0   
3     Renault        Laguna  2012  Universal  110000.0                1500.0   
4        Opel         Astra  2006  Universal  200000.0                1600.0   
5    Mercedes          Vito  2000   Microvan  300000.0                2200.0   

      Fuel_type Transmission  Price(euro)  
0        Hybrid    Automatic       7750.0  
1        Diesel       Manual       8550.0  
2        Petrol       Manual       2200.0  
3        Diesel       Manual       6550.0  
4  Metan/Propan       Manual       4100.0  
5        Diesel       Manual       3490.0

Метод .tail(n) - последние n строк

In [ ]:
DF.tail(9) # то же самое только последние строки 

Make         Model  Year      Style  Distance  \
37255     Renault        Megane  2011      Sedan  145000.0   
37256      Toyota         Auris  2007  Hatchback  222111.0   
37257       Skoda        Superb  2016  Cabriolet   78900.0   
37258  Volkswagen        Passat  2016      Sedan   88000.0   
37259  Land Rover    Freelander  2002  Crossover  225000.0   
37260       Dacia     Logan Mcv  2015  Universal   89000.0   
37261     Renault         Modus  2009  Hatchback     225.0   
37262       Mazda             6  2006      Combi  370000.0   
37263     Renault  Grand Scenic  2006    Minivan  300000.0   

       Engine_capacity(cm3)     Fuel_type Transmission  Price(euro)  
37255                1500.0        Diesel    Automatic       6200.0  
37256                1400.0        Diesel    Automatic       4444.0  
37257                2000.0        Petrol    Automatic      21000.0  
37258                1800.0        Petrol    Automatic      11500.0  
37259                1800.0  Metan/Propan       Manual       4400.0  
37260                1500.0        Diesel       Manual       7000.0  
37261                1500.0        Diesel       Manual       4500.0  
37262                2000.0        Diesel       Manual       4000.0  
37263                1500.0        Diesel       Manual       4000.0

**Индексация**

аттрибут .loc[строки, столбцы] 

In [ ]:
DF.loc[1217,'Fuel_type'] #  смотрим показатель конкретного показателя на пересечении строки столбаца 

'Diesel'

In [ ]:
DF.loc[56]#  смотрим что у нас на строке номер 56

Make                          BMW
Model                          X5
Year                         2005
Style                         SUV
Distance                 260500.0
Engine_capacity(cm3)       3000.0
Fuel_type                  Petrol
Transmission            Automatic
Price(euro)                8999.0
Name: 56, dtype: object

In [ ]:
DF.loc[:,'Style']# смотрим что на столбце  Style 

0        Hatchback
1        Universal
2        Hatchback
3        Universal
4        Universal
           ...    
37259    Crossover
37260    Universal
37261    Hatchback
37262        Combi
37263      Minivan
Name: Style, Length: 37264, dtype: object

In [ ]:
DF.loc[:,['Style']].head(5) #  смотрим тоже самое, только верхние 5 покзателей 

Style
0  Hatchback
1  Universal
2  Hatchback
3  Universal
4  Universal

In [ ]:
DF.loc[:,['Price(euro)','Year']].head(5) #  смотрим 2 столбца 

Price(euro)  Year
0       7750.0  2011
1       8550.0  2014
2       2200.0  1998
3       6550.0  2012
4       4100.0  2006

In [ ]:
DF.loc[:,'Make':'Style'].head(5) #. смотрим все столбца между заданными включительно их 

Make         Model  Year      Style
0      Toyota         Prius  2011  Hatchback
1     Renault  Grand Scenic  2014  Universal
2  Volkswagen          Golf  1998  Hatchback
3     Renault        Laguna  2012  Universal
4        Opel         Astra  2006  Universal

In [ ]:
DF.loc[69,:]

Make                        Honda
Model                        CR-V
Year                         2002
Style                   Universal
Distance                 275100.0
Engine_capacity(cm3)       2000.0
Fuel_type                  Petrol
Transmission               Manual
Price(euro)                5900.0
Name: 69, dtype: object

In [ ]:
DF.loc[69:69,:]

Make Model  Year      Style  Distance  Engine_capacity(cm3) Fuel_type  \
69  Honda  CR-V  2002  Universal  275100.0                2000.0    Petrol   

   Transmission  Price(euro)  
69       Manual       5900.0

In [ ]:
DF.loc[322:1437,:]

Make     Model    Year      Style  Distance Engine_capacity(cm3)  \
322      Skoda   Octavia  2010.0  Universal       1.0               1400.0   
323   Mercedes   E Class  2016.0      Sedan  161024.0               2200.0   
324      Honda     Civic  2006.0  Hatchback  195000.0               1400.0   
325       Ford    Fusion  2015.0      Sedan  124000.0               2000.0   
326       Ford    Fusion  2003.0  Hatchback  330000.0               1400.0   
...        ...       ...     ...        ...       ...                  ...   
1433   Peugeot       208  2013.0  Hatchback  154000.0               1400.0   
1434  Infiniti      QX30  2017.0        SUV   91000.0               1500.0   
1435      Ford      Kuga  2016.0        SUV  102500.0               1500.0   
1436      Ford      Kuga  2018.0        SUV  122100.0               1500.0   
1437       BMW  3 Series  2016.0      Sedan   64000.0               2000.0   

           Fuel_type Transmission  Price(euro)  
322           Petrol       Manual       5350.0  
323           Diesel    Automatic      26950.0  
324           Petrol       Manual       5400.0  
325   Plug-in Hybrid    Automatic      11499.0  
326           Diesel    Automatic       4100.0  
...              ...          ...          ...  
1433          Diesel    Automatic       6999.0  
1434          Diesel    Automatic      15490.0  
1435          Diesel       Manual      14490.0  
1436          Diesel       Manual      14250.0  
1437          Petrol    Automatic      17500.0  

[1116 rows x 9 columns]

In [ ]:
DF.loc[227:229,'Make':'Fuel_type']

Make    Model  Year      Style  Distance  Engine_capacity(cm3)  \
227     Opel  Calibra  1992      Coupe  400000.0                2000.0   
228  Renault   Kadjar  2016  Crossover  150425.0                1500.0   
229     Ford   Fusion  2018      Sedan  105000.0                2000.0   

        Fuel_type  
227  Metan/Propan  
228        Diesel  
229        Hybrid

**Логическое Индексирование**

In [ ]:
DF[DF['Transmission']=='Manual']

Make         Model  Year      Style  Distance  \
1         Renault  Grand Scenic  2014  Universal  135000.0   
2      Volkswagen          Golf  1998  Hatchback       1.0   
3         Renault        Laguna  2012  Universal  110000.0   
4            Opel         Astra  2006  Universal  200000.0   
5        Mercedes          Vito  2000   Microvan  300000.0   
...           ...           ...   ...        ...       ...   
37259  Land Rover    Freelander  2002  Crossover  225000.0   
37260       Dacia     Logan Mcv  2015  Universal   89000.0   
37261     Renault         Modus  2009  Hatchback     225.0   
37262       Mazda             6  2006      Combi  370000.0   
37263     Renault  Grand Scenic  2006    Minivan  300000.0   

       Engine_capacity(cm3)     Fuel_type Transmission  Price(euro)  
1                    1500.0        Diesel       Manual       8550.0  
2                    1400.0        Petrol       Manual       2200.0  
3                    1500.0        Diesel       Manual       6550.0  
4                    1600.0  Metan/Propan       Manual       4100.0  
5                    2200.0        Diesel       Manual       3490.0  
...                     ...           ...          ...          ...  
37259                1800.0  Metan/Propan       Manual       4400.0  
37260                1500.0        Diesel       Manual       7000.0  
37261                1500.0        Diesel       Manual       4500.0  
37262                2000.0        Diesel       Manual       4000.0  
37263                1500.0        Diesel       Manual       4000.0  

[20723 rows x 9 columns]

In [ ]:
DF[(DF['Transmission']=='Manual')*(DF['Price(euro)']>5000)]

Make          Model  Year      Style  Distance  \
1         Renault   Grand Scenic  2014  Universal  135000.0   
3         Renault         Laguna  2012  Universal  110000.0   
6      Volkswagen         Passat  2010  Universal  290000.0   
15     Volkswagen      Golf Plus  2010  Hatchback  150000.0   
17         Nissan        Qashqai  2011  Crossover  251000.0   
...           ...            ...   ...        ...       ...   
37246  Volkswagen      Golf Plus  2007      Sedan  186000.0   
37248  Volkswagen         Passat  2016  Universal  180000.0   
37253        Opel          Astra  2012  Universal  180000.0   
37254      Toyota  Avensis Verso  2004    Minivan  245627.0   
37260       Dacia      Logan Mcv  2015  Universal   89000.0   

       Engine_capacity(cm3) Fuel_type Transmission  Price(euro)  
1                    1500.0    Diesel       Manual       8550.0  
3                    1500.0    Diesel       Manual       6550.0  
6                    1390.0    Petrol       Manual       5600.0  
15                   1200.0    Petrol       Manual       5500.0  
17                   1500.0    Diesel       Manual       8300.0  
...                     ...       ...          ...          ...  
37246                2000.0    Diesel       Manual       7000.0  
37248                1800.0    Diesel       Manual      11500.0  
37253                1300.0    Diesel       Manual       6400.0  
37254                2000.0    Diesel       Manual       7490.0  
37260                1500.0    Diesel       Manual       7000.0  

[8294 rows x 9 columns]

# Сортировка DataFrame



Метод .sort_values(by)

In [ ]:
DF.sort_values(by = 'Price(euro)').tail(5) # сортируем по цене, берем самые дорогие 

Make               Model  Year      Style  Distance  \
13131      Toyota  Land Cruiser Prado  2012        SUV   10005.0   
14834  Volkswagen               Caddy  2015      Combi  158260.0   
29132    Mercedes             G Class  2020        SUV    3000.0   
17849         KIA            Sportage  2018  Crossover   46000.0   
33677  Brilliance                 BS3  2009      Sedan   57000.0   

       Engine_capacity(cm3)     Fuel_type Transmission  Price(euro)  
13131                3000.0        Diesel    Automatic     200000.0  
14834                1395.0  Metan/Propan       Manual     200000.0  
29132                3982.0        Petrol    Automatic     215000.0  
17849                1600.0        Petrol    Automatic     490000.0  
33677                1598.0        Petrol       Manual   10000000.0


Метод .sort_values(by, ascending= False)

In [ ]:
DF.sort_values(by = 'Year', ascending= False).head(5) #. сортируем по году производства, смотрим самые новые 

Make      Model  Year      Style  Distance  Engine_capacity(cm3)  \
29164        Audi         A5  2021      Sedan     661.0                2000.0   
32583    Mercedes  GLS Klass  2021        SUV       0.0                3000.0   
15611  Volkswagen       Polo  2021  Universal  280000.0                1400.0   
6044        Skoda     Superb  2021      Sedan       0.0                1500.0   
11840      Toyota    Corolla  2021  Universal     288.0                1987.0   

      Fuel_type Transmission  Price(euro)  
29164    Petrol    Automatic      32200.0  
32583    Diesel    Automatic     107528.0  
15611    Petrol       Manual       1900.0  
6044     Petrol    Automatic      30000.0  
11840    Hybrid    Automatic      23000.0

In [ ]:
DF.loc[69:322,:].sort_values(by = "Distance", ascending= False)

Make      Model  Year      Style    Distance  Engine_capacity(cm3)  \
281     Renault     Megane  2004  Hatchback  79593986.0                1600.0   
138     Citroen   Berlingo  2001    Minivan  68286164.0                1900.0   
223  Volkswagen  Passat CC  1994      Sedan  68204075.0                1800.0   
218        Opel     Zafira  2002    Minivan  11111111.0                2000.0   
276         Vaz   2121niva  1982  Hatchback   1563215.0                1600.0   
..          ...        ...   ...        ...         ...                   ...   
305        Audi         A6  2018      Sedan         0.0                1800.0   
234      Toyota      Prius  2016  Hatchback         0.0                1800.0   
238       Dacia      Lodgy  2013  Universal         0.0                1500.0   
192         KIA    Sorento  2017  Crossover         0.0                2200.0   
259        Opel     Meriva  2005  Hatchback         0.0                1400.0   

        Fuel_type Transmission  Price(euro)  
281        Petrol       Manual       1050.0  
138        Diesel       Manual       2050.0  
223        Petrol       Manual       1550.0  
218  Metan/Propan       Manual       3700.0  
276        Petrol       Manual       1050.0  
..            ...          ...          ...  
305        Petrol    Automatic      25999.0  
234        Hybrid    Automatic      12350.0  
238        Diesel       Manual       7490.0  
192        Diesel    Automatic      21490.0  
259        Petrol       Manual       3550.0  

[254 rows x 9 columns]

In [ ]:
DF

Make         Model  Year      Style  Distance  \
0          Toyota         Prius  2011  Hatchback  195000.0   
1         Renault  Grand Scenic  2014  Universal  135000.0   
2      Volkswagen          Golf  1998  Hatchback       1.0   
3         Renault        Laguna  2012  Universal  110000.0   
4            Opel         Astra  2006  Universal  200000.0   
...           ...           ...   ...        ...       ...   
37259  Land Rover    Freelander  2002  Crossover  225000.0   
37260       Dacia     Logan Mcv  2015  Universal   89000.0   
37261     Renault         Modus  2009  Hatchback     225.0   
37262       Mazda             6  2006      Combi  370000.0   
37263     Renault  Grand Scenic  2006    Minivan  300000.0   

       Engine_capacity(cm3)     Fuel_type Transmission  Price(euro)  
0                    1800.0        Hybrid    Automatic       7750.0  
1                    1500.0        Diesel       Manual       8550.0  
2                    1400.0        Petrol       Manual       2200.0  
3                    1500.0        Diesel       Manual       6550.0  
4                    1600.0  Metan/Propan       Manual       4100.0  
...                     ...           ...          ...          ...  
37259                1800.0  Metan/Propan       Manual       4400.0  
37260                1500.0        Diesel       Manual       7000.0  
37261                1500.0        Diesel       Manual       4500.0  
37262                2000.0        Diesel       Manual       4000.0  
37263                1500.0        Diesel       Manual       4000.0  

[37264 rows x 9 columns]

# Сохранение DataFrame в файл


Метод .to_csv(path, index) 

In [39]:
DF.to_csv('/content/cars_moldova_no_dup.csv',index=False) #. сохраняем все в файл 

# Контрольные вопросы

Вопрос: 
<b> Назовите тип данных категориальных столбцов в Pandas


object, int64, float64

Вопрос: 
<b>Назовите целевой столбец для задачи регрессии

Вопрос: 
<b>Назовите целевой столбец для задачи классификации


Style

Вопрос: 
<b>Возьмите датафрейм после удаления дубликатов. Выберете из полного датафрейма строки с индекса 69 по 322. Отсортируйте полученный датафрейм по колонке  'Distance' по убыванию. Какое значение колонки  'Style' у полученного датафрейма во второй строке сверху 

элемент 298, Style - Sedan 

Вопрос:
<b> Укажите правильную размерность исходного набора данных

2.8 mb, 2740252 bytes